blockchain/headers.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (C) 2021 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package blockchain
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/spacemonkeygo/monkit/v3"
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+)
+
+var (
+	// ErrNoHeader is error thrown when there is no header in db.
+	ErrNoHeader = errs.New("HeadersDB: header not found")
+	// ErrHeaderCache is headers cache error class.
+	ErrHeaderCache = errs.Class("HeadersCache")
+
+	mon = monkit.Package()
+)
+
+// Header holds ethereum blockchain block header indexed data.
+// No need to keep number as big.Int right now as block count on ethereum mainnet is far from overflowing int64 capacity.
+type Header struct {
+	Hash      Hash
+	Number    int64
+	Timestamp time.Time
+}
+
+// HeadersDB is ethereum blockchain block header indexed cache.
+//
+// architecture: Database
+type HeadersDB interface {
+	// Insert inserts new header to cache db.
+	Insert(ctx context.Context, hash Hash, number int64, timestamp time.Time) error
+	// Delete deletes header from db by hash.
+	Delete(ctx context.Context, hash Hash) error
+	// Get retrieves header by hash.
+	Get(ctx context.Context, hash Hash) (Header, error)
+	// GetByNumber retrieves header by number.
+	GetByNumber(ctx context.Context, number int64) (Header, error)
+	// List retrieves all headers stored in cache db.
+	List(ctx context.Context) ([]Header, error)
+}
+
+// HeadersCache is a blockchain block header cache.
+type HeadersCache struct {
+	log *zap.Logger
+	db  HeadersDB
+}
+
+// NewHeadersCache creates new header cache.
+func NewHeadersCache(log *zap.Logger, db HeadersDB) *HeadersCache {
+	return &HeadersCache{
+		log: log,
+		db:  db,
+	}
+}
+
+// Get retrieves block header from cache by its hash.
+func (headersCache *HeadersCache) Get(ctx context.Context, hash Hash) (Header, bool, error) {
+	header, err := headersCache.db.Get(ctx, hash)
+	if err != nil {
+		if errs.Is(err, ErrNoHeader) {
+			return Header{}, false, nil
+		}
+
+		return Header{}, false, ErrHeaderCache.Wrap(err)
+	}
+
+	return header, true, nil
+}
+
+// HeaderSearch is block header search context.
+type HeaderSearch struct {
+	log         *zap.Logger
+	endpoint    string
+	blockTime   time.Duration
+	threshold   time.Duration
+	batchLimit  int64
+	searchLimit int
+}
+
+// NewHeaderSearch creates new block header search context.
+func NewHeaderSearch(log *zap.Logger, endpoint string, blockTime, threshold time.Duration, batchLimit int64, searchLimit int) *HeaderSearch {
+	return &HeaderSearch{
+		log:         log,
+		endpoint:    endpoint,
+		blockTime:   blockTime,
+		threshold:   threshold,
+		batchLimit:  batchLimit,
+		searchLimit: searchLimit,
+	}
+}
+
+// After finds first block after specified timestamp and returns its header.
+func (headerSearch *HeaderSearch) After(ctx context.Context, t time.Time) (_ Header, err error) {
```

Comment:
Ok, from an offline discussion, I learned:

 1. It might be slightly better than the binary search if the block time can be estimated in advance


Question: is the block time predictable enough on polygon (and avalanche)?
 
Also learned that this method is for a different use-case: getting the transactions of a certain period of time: I was confused because it's not exactly the same goal as adding date to each of the log events (separating them to a different PR may help the understanding)